Tests/test_image_filter.py

```diff
@@ -5,90 +5,110 @@
+def test_crash(size):
+    im = Image.new("RGB", size)
     im.filter(ImageFilter.SMOOTH)
 
 
-def test_modefilter():
-    def modefilter(mode):
-        im = Image.new(mode, (3, 3), None)
-        im.putdata(list(range(9)))
-        # image is:
-        #   0 1 2
-        #   3 4 5
-        #   6 7 8
-        mod = im.filter(ImageFilter.ModeFilter).getpixel((1, 1))
-        im.putdata([0, 0, 1, 2, 5, 1, 5, 2, 0])  # mode=0
-        mod2 = im.filter(ImageFilter.ModeFilter).getpixel((1, 1))
-        return mod, mod2
-
-    assert modefilter("1") == (4, 0)
-    assert modefilter("L") == (4, 0)
-    assert modefilter("P") == (4, 0)
-    assert modefilter("RGB") == ((4, 0, 0), (0, 0, 0))
-
-
-def test_rankfilter():
-    def rankfilter(mode):
-        im = Image.new(mode, (3, 3), None)
+@pytest.mark.parametrize(
+    "mode,expected",
+    (
+        ("1", (4, 0)),
+        ("L", (4, 0)),
+        ("P", (4, 0)),
+        ("RGB", ((4, 0, 0), (0, 0, 0))),
+    ),
+)
+def test_modefilter(mode, expected):
+    im = Image.new(mode, (3, 3), None)
+    im.putdata(list(range(9)))
+    # image is:
+    #   0 1 2
+    #   3 4 5
+    #   6 7 8
+    mod = im.filter(ImageFilter.ModeFilter).getpixel((1, 1))
+    im.putdata([0, 0, 1, 2, 5, 1, 5, 2, 0])  # mode=0
+    mod2 = im.filter(ImageFilter.ModeFilter).getpixel((1, 1))
+    assert (mod, mod2) == expected
+
+
+@pytest.mark.parametrize(
+    "mode,expected",
+    (
+        ("1", (0, 4, 8)),
+        ("L", (0, 4, 8)),
+        ("RGB", ((0, 0, 0), (4, 0, 0), (8, 0, 0))),
+        ("I", (0, 4, 8)),
+        ("F", (0.0, 4.0, 8.0)),
+    ),
+)
+def test_rankfilter(mode, expected):
+    im = Image.new(mode, (3, 3), None)
+    im.putdata(list(range(9)))
+    # image is:
+    #   0 1 2
+    #   3 4 5
+    #   6 7 8
+    minimum = im.filter(ImageFilter.MinFilter).getpixel((1, 1))
+    med = im.filter(ImageFilter.MedianFilter).getpixel((1, 1))
+    maximum = im.filter(ImageFilter.MaxFilter).getpixel((1, 1))
+    assert (minimum, med, maximum) == expected
+
+
+def test_rankfilter_error():
+    with pytest.raises(ValueError):
+        im = Image.new("P", (3, 3), None)
         im.putdata(list(range(9)))
         # image is:
         #   0 1 2
         #   3 4 5
         #   6 7 8
-        minimum = im.filter(ImageFilter.MinFilter).getpixel((1, 1))
-        med = im.filter(ImageFilter.MedianFilter).getpixel((1, 1))
-        maximum = im.filter(ImageFilter.MaxFilter).getpixel((1, 1))
-        return minimum, med, maximum
-
-    assert rankfilter("1") == (0, 4, 8)
-    assert rankfilter("L") == (0, 4, 8)
-    with pytest.raises(ValueError):
-        rankfilter("P")
-    assert rankfilter("RGB") == ((0, 0, 0), (4, 0, 0), (8, 0, 0))
-    assert rankfilter("I") == (0, 4, 8)
-    assert rankfilter("F") == (0.0, 4.0, 8.0)
+        im.filter(ImageFilter.MinFilter).getpixel((1, 1))
+        im.filter(ImageFilter.MedianFilter).getpixel((1, 1))
+        im.filter(ImageFilter.MaxFilter).getpixel((1, 1))
```

Comment:
The second and third filter are not tested, I'd parametrize this by the image filter.

Reply:
resolved with 65694f3fb82bd6b29f1b8750f730ba311e41f8e5